clang/test/CodeGenSYCL/regcall-cc-test.cpp

```diff
@@ -0,0 +1,360 @@
+// CHECK-DAG: define weak_odr x86_regcallcc i32 @_Z16__regcall3__funcI10ShortShortET_S1_(i32 %{{[0-9a-zA-Z_.]+}})
+
+struct CharShort {
+  char a;
+  short b;
+
+  CharShort foo() { return *this; }
+};
+// CHECK-DAG: %struct.CharShort = type { i8, i16 }
+
+template SYCL_DEVICE CharShort __regcall func<CharShort>(CharShort x);
+// CHECK-DAG: define weak_odr x86_regcallcc i32 @_Z16__regcall3__funcI9CharShortET_S1_(i32 %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with primitive single field element is just unwrapped
+
+struct Char {
+  char a;
+
+  Char foo() { return *this; }
+};
+// CHECK-DAG: %struct.Char = type { i8 }
+
+template SYCL_DEVICE Char __regcall func<Char>(Char x);
+// CHECK-DAG: define weak_odr x86_regcallcc i8 @_Z16__regcall3__funcI4CharET_S1_(i8 %{{[0-9a-zA-Z_.]+}})
+
+struct Float {
+  float a;
+
+  Float foo() { return *this; }
+};
+// CHECK-DAG: %struct.Float = type { float }
+
+template SYCL_DEVICE Float __regcall func<Float>(Float x);
+// CHECK-DAG: define weak_odr x86_regcallcc float @_Z16__regcall3__funcI5FloatET_S1_(float %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with sizeof() == 5-8 is passed as two 32-bit integers
+// 32-bit integers
+
+struct CharCharShortFloat {
+  char a, b;
+  short c;
+  float d;
+
+  CharCharShortFloat foo() { return *this; }
+};
+// CHECK-DAG: %struct.CharCharShortFloat = type { i8, i8, i16, float }
+
+template SYCL_DEVICE CharCharShortFloat __regcall func<CharCharShortFloat>(CharCharShortFloat x);
+// CHECK-DAG: define weak_odr x86_regcallcc [2 x i32] @_Z16__regcall3__funcI18CharCharShortFloatET_S1_([2 x i32] %{{[0-9a-zA-Z_.]+}})
+
+// === TEST CASE: a struct with some of the fields padded and sizeof() > 8
+//   * when passed as argument, it is broken into constituents
+//   * is returned by value
+
+struct CharFloatCharShort {
+  char a;
+  float b;
+  char c;
+  short d;
+
+  CharFloatCharShort foo() { return *this; }
+};
+
+// CHECK-DAG: %struct.CharFloatCharShort = type { i8, float, i8, i16 }
+
+template SYCL_DEVICE CharFloatCharShort __regcall func<CharFloatCharShort>(CharFloatCharShort x);
+// CHECK-DAG: define weak_odr x86_regcallcc %struct.CharFloatCharShort @_Z16__regcall3__funcI18CharFloatCharShortET_S1_(i8 %{{[0-9a-zA-Z_.]+}}, float %{{[0-9a-zA-Z_.]+}}, i8 %{{[0-9a-zA-Z_.]+}}, i16 %{{[0-9a-zA-Z_.]+}})
+
+struct CharDoubleCharLonglong {
+  char a;
+  double b;
+  char c;
+  long long d;
+
+  CharDoubleCharLonglong foo() { return *this; }
+};
+
+// CHECK-DAG: %struct.CharDoubleCharLonglong = type { i8, double, i8, i64 }
+
+template SYCL_DEVICE CharDoubleCharLonglong __regcall func<CharDoubleCharLonglong>(CharDoubleCharLonglong x);
+// CHECK-DAG: define weak_odr x86_regcallcc %struct.CharDoubleCharLonglong @_Z16__regcall3__funcI22CharDoubleCharLonglongET_S1_(i8 %{{[0-9a-zA-Z_.]+}}, double %{{[0-9a-zA-Z_.]+}}, i8 %{{[0-9a-zA-Z_.]+}}, i64 %{{[0-9a-zA-Z_.]+}})
+
+
+// === TEST CASE: a struct of 130x4-byte elements is still passed by value
```

Comment:
Could you add a comment why you chose 130?   Does it hit some threshold over which something different happens?

Reply:
This is just "arbitrarily big" number, more than 128 to test big number of fields does not prevent passing by values.
I'll add a comment.